[6.x] Add cache-control header to cached content

indykoning opened the pull request:
Closes https://github.com/statamic/ideas/issues/1309

Half and full caching strategies will get these improvements if these are pages that can be cached.

This PR will add cache-control public headers for pages that can be/will be/have been cached or have been served from cache if the webserver wasn't configured properly or query params did not match.

It can improve experienced speed and reduce server load by using local copies of statically cached content. And if combined with a properly configured CDN which supports caching pages can reduce server load as the CDN will serve the page and occasionally update it's cache.

~~(If you're using a CDN in-between you might not even need the full strategy anymore)~~
Edit: ^ That is not true (yet) since it still requires the csrf token + session, cookies will be sent. Which means CDNs will not cache unless specifically instructed to ignore the cookies

indykoning committed:
Add cache-control header to cached content

duncanmcclean commented on src/StaticCaching/ServiceProvider.php:
```diff
@@ -91,6 +91,14 @@ public function boot()
             return app(Cacher::class)->getUrl($this);
         });
 
+        Response::macro('makeCacheControlCacheable', function () {
```
Looks like there's a missing import here, causing the test suite to fail. Can you fix it?

indykoning replied:
Fixed, during development i had it in the ResponseStatusTracker which was not the right place in the end but had the import 😅

indykoning committed:
Added request import

indykoning committed:
Removed duplicate response import

indykoning committed:
Fix linting

indykoning committed:
Update config

indykoning commented:
This PR should now be ready to merge now, it is currently in use without any issues 🙂

indykoning committed:
Run isNotModified check

Calling this check triggers the ETag matching, and sets the notModified response code if a match is found. 
If https://github.com/laravel/framework/blob/12.x/src/Illuminate/Http/Middleware/CheckResponseForModifications.php is in the middleware stack it would do it too, it is not in the stack by default. And checking multiple times causes no issues

indykoning committed:
Check getContent before setting etag

indykoning committed:
Fix linting

indykoning committed:
Added check to only execute on macrod response

indykoning committed:
Merge branch '5.x' into feature/cache-control-header

indykoning committed:
Merge branch '6.x' into feature/cache-control-header

indykoning committed:
Check if cacher is FileCacher before caching response

indykoning commented:
Seeing the way half cache operates i've decided it's the best choice to hold off on that for now, and focus on using it for full page cache for the current scope.

Since half strategy cache replaces the CSRF token and nocache tags before serving the response, nocache content will most-likely get cached. we could work around it by using a shared max age of 0. meaning your browser will get allowed to cache, but your CDN/Proxies do not.
Still it could cause issues if your max age is set too high.

This is of no issue for full strategy as we're already considering that the entire file will be statically cached and no replacements can be done server-side. So all replacements for nocache and CSRF tokens are done client-side.

I've updated the PR to focus on the full strategy
